[core] Parallel run and string sort order

danielpoltx opened the pull request:
Hi !

For more information see issue #664 

This adds:
1. YCSB uses an integer sort order while Hbase use region splits based on string sort order. So for Hbase “user1000” comes before “user2” and messes up region pre-splits. I've added a zeropadding parameter to specify the width of the record number. Defaults to 1
2. YCSB can split the load across multiple instance, but not the run.Added insertstart and insertcount to run. The functionality is similar to load. This is great for targeting specific areas of your table or troubleshooting.
3. Introduced an “ordered” distribution. Useful in debug situations.

Have a nice day,
Dani

kruthar commented:
@danielpoltx thanks for the work you have done here, at this point I and who ever else wishes to review the code will post any necessary fixes.

Also, for future reference it is advisable to put each separate feature in it's own PR, that way change sets are small and easy to review.

kruthar commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -452,8 +473,10 @@ public void init(Properties p) throws WorkloadException {
-      keychooser = new UniformIntegerGenerator(0, recordcount - 1);
-    } else if (requestdistrib.compareTo("zipfian") == 0) {
+      keychooser=new UniformIntegerGenerator(insertstart,insertstart+insertcount-1);
+    } else if (requestdistrib.compareTo("ordered")==0)	{
```
Should follow the convention here and actually name expect "sequential" here since you are using the sequential generator.

danielpoltx replied:
I tried to reuse the "ordered" keyword from insertorder. I can either change the sequential generator to "OrderedGenerator" or change "ordered" keyword to "sequential" keyword. Do you prefer one vs the other ?

saintstack replied:
Sequential seems more descriptive of what is going on IMO

danielpoltx replied:
Ok, I'll change the name to Sequential

danielpoltx replied:
Done

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -62,10 +62,16 @@
  * modify it, write it back (default: 0)
  * <LI><b>requestdistribution</b>: what distribution should be used to select the records to operate
- * on - uniform, zipfian, hotspot, ordered or latest (default: uniform)
+ * on - uniform, zipfian, hotspot, ordered, exponential or latest (default: uniform)
```
Is this an unrelated doc fix?

danielpoltx replied:
Yes, I've noticed in the code there's also an exponential distribution. Not my doing

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -62,10 +62,16 @@
  * number of records to scan, for each scan, between 1 and maxscanlength (default: uniform)
+ * <LI><b>insertstart</b>: for parallel loads and runs, defines the starting record for this
+ * YCSB instance (default: 0)
+ * <LI><b>insertcount</b>: for parallel loads and runs, defines the ending record for this
```
So, an instance will start at 'insertstart' record and go for 'insertcount' inserts? Is 'ending record' the right way to describe this? Its not exactly the ending record... but a count? This is a nit.

danielpoltx replied:
You are correct. I'll update the doc

danielpoltx replied:
Done

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -62,10 +62,16 @@
+ * YCSB instance (default: 0)
+ * <LI><b>insertcount</b>: for parallel loads and runs, defines the ending record for this
+ * YCSB instance (default: recordcount)
+ * <LI><b>zeropadding</b>: for generating a record sequence compatible with string sort order by
```
Suggest you add an example of what padding will do here. Is a default of '1' any good? Should it be 8 or something for any decent workload?

danielpoltx replied:
Ok, I'll add example. Default=1 is basically no padding which is current behavior. Didn't want to mess up the existing environment for people upgrading to a new YCSB version.

saintstack replied:
Oh. That makes sense. Leave it as is.

danielpoltx replied:
Done

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -505,7 +525,7 @@ public String buildKeyName(long keynum) {
       keynum = Utils.hash(keynum);
     }
-    return "user" + keynum;
+	return "user"+String.format("%0"+zeropadding+"d",keynum);
```
Just to note that format is costly. Probably not important in the scheme of things.

danielpoltx replied:
Any suggesting how to replace it with something less costly that has the same result in our case ?

saintstack replied:
I think it is fine for now but if you want to dig more, check out https://issues.apache.org/jira/browse/HBASE-15569. Manually doing the substitution was more performant.

danielpoltx replied:
Done. Testing showed format was ~11x slower compared to original code (no padding) and this new version is only ~20% slower compared to original code, while it does padding.

risdenk replied:
Not sure it makes sense but can we short circuit if there is no padding with an if statement? That would avoid a slowdown compared to the original code?

danielpoltx replied:
I'm new to Github/coding so I don't know how to make this comment look at the latest code I pushed to the branch, but there's the latest function that addresses performance concerns
public String buildKeyName(long keynum) {
    if (!orderedinserts) {
      keynum = Utils.hash(keynum);
    }
	String value=Long.toString(keynum);
	int fill=zeropadding - value.length();
	String prekey="user";
	for(int i=0; i<fill;i++) {prekey+='0';}
	return prekey+value;
  }

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -393,8 +394,8 @@ public void init(Properties p) throws WorkloadException {
 
-    int insertstart =
-        Integer.parseInt(p.getProperty(INSERT_START_PROPERTY, INSERT_START_PROPERTY_DEFAULT));
+	int insertstart=Integer.parseInt(p.getProperty(INSERT_START_PROPERTY,INSERT_START_PROPERTY_DEFAULT));
```
Retain the original formatting Daniel... there is probably a reason (line length or something).

danielpoltx replied:
okidoki.

danielpoltx replied:
Done

risdenk commented:
@danielpoltx the Travis build failed with the following:

```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/travis/build/brianfrankcooper/YCSB/core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:[35,32] cannot find symbol
  symbol:   class SequentialGenerator
  location: package com.yahoo.ycsb.generator
[ERROR] /home/travis/build/brianfrankcooper/YCSB/core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:[478,40] cannot find symbol
  symbol:   class SequentialGenerator
  location: class com.yahoo.ycsb.workloads.CoreWorkload
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.3:compile (default-compile) on project core: Compilation failure: Compilation failure:
[ERROR] /home/travis/build/brianfrankcooper/YCSB/core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:[35,32] cannot find symbol
[ERROR] symbol:   class SequentialGenerator
[ERROR] location: package com.yahoo.ycsb.generator
[ERROR] /home/travis/build/brianfrankcooper/YCSB/core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:[478,40] cannot find symbol
[ERROR] symbol:   class SequentialGenerator
[ERROR] location: class com.yahoo.ycsb.workloads.CoreWorkload
[ERROR] -> [Help 1]
```

saintstack commented:
Patch looks good. I tried the new args.

I did this and stuff it hung:

`$ ./bin/ycsb run hbase10  -P ./workloads/workloadc -p table=ycsb -cp ~/conf_hbase -p columnfamily=family -threads 1 -p recordcount=10 -p operationcount=10  -p insertstart=100 -p insertcount=2 -p zeropadding=5`

I think having insertstart > operationcount is a problem?

...but this worked nicely:
`
$ ./bin/ycsb run hbase10  -P ./workloads/workloadc -p table=ycsb -cp ~/conf_hbase -p columnfamily=family -threads 1 -p recordcount=5 -p operationcount=5  -p insertstart=2 -p insertcount=2 -p zeropadding=5`

... though I'm not sure the zeropadding is working?

```
hbase(main):001:0> scan 'ycsb2'
ROW                                                                                                                              COLUMN+CELL
 user01820151046732198393                                                                                                        column=family:field0, timestamp=1460701028309, value=%=4 E)3:p+Pg)9(8<8<_y?*n,2~7+,%@)(32,6x75l82h&L=;Lk&/4(5z.Vs,I=5/z!K3$?,85 9=l?,*8@a"Ma33`5)412t)C1!
 user01820151046732198393                                                                                                        column=family:field1, timestamp=1460701028309, value=-He6>><\x5Cu2D-1[9(*d&X}#5*/Y+ (j,S',+j9;x3"b;Z%0(x3)d2K%<7`#Q)%_)%)8:A#)R!!E5=)4;D753r6=>1)$>'*)?,*9 4
 user01820151046732198393                                                                                                        column=family:field2, timestamp=1460701028309, value=#-:5D#4Eo.E?<S\x7F"U+%Sk*6 ?!l3M%9(<,:6;B7(?>(Q%92b24j"K92.z>$$5*z*T{!:r"3p56((929Ok'Le+Ew:22<B-4]#"2r/
 user01820151046732198393                                                                                                        column=family:field3, timestamp=1460701028309, value=:/d2/2+M+8V!#4x2M/(4<9Wk!=|6Oe": =)<5S+((0'A7)Ci:F+1?z%<|?!`3829)z 925Cy+8j3U{=@/(R39]',[w=Ii;^9<R;0
 user01820151046732198393                                                                                                        column=family:field4, timestamp=1460701028309, value= Z?,>>()>)Ps;5d/'6(7,!:f7Ky?5<)."%Y;4$j&,<786)7j68.("47:$4+p Z'(U3<-$-Le086,&4&:f;)p9_o:&r+4*8Uo!R%9
 user01820151046732198393                                                                                                        column=family:field5, timestamp=1460701028309, value=(C#,!z'3f>E5&7,4?*25`+!"0T3-Zy=\x5Cs'<"&@q:^s'Tg:Uc*Dq?-p4W#,M}.F/$;2+6.!/"+S?8Ik$+l'?h-Me/F19\x5Ce#4h;V18
 user01820151046732198393                                                                                                        column=family:field6, timestamp=1460701028309, value=4:*8:<*W)(E;9_3$V301$98z%6|;5 ;:.!*l2#~*"f-R9(D+ L==4l=M=(B%$2,%"2;]k.H7!M)0:x)6j#T5&.x?%x"V7 1h?$h"
 user01820151046732198393                                                                                                        column=family:field7, timestamp=1460701028309, value=(001Q+#&  #f!*l&B/9Fq?V#0A3!5f?M15_s0>z2:j:\x5C!4\x5Ce):$$W;.0b*/2%Wo.E--_!9^}:"f2Hg(:t7Ds+Ry&>d!Z'8>d6#4,
 user01820151046732198393                                                                                                        column=family:field8, timestamp=1460701028309, value=<F\x7F?Ps?E=,Ak)D9!Ec7.`"6(0N78.|4Nq>) 9Ng+X18^y;L#7= ,D\x7F&@9=&,.S/%640-r38,=K',Pc(>*=Ik"Mk&R{>!b7?40]?8
 user01820151046732198393                                                                                                        column=family:field9, timestamp=1460701028309, value=#1n?-n7..!I!8;($Y1-58(.b3_1!5h;^m#! :;4<:v#C31)(4=>!..$@70,r8V}6Zm=Tw6$x+"64!(2.(0)81"l2Ca#%4<'.;*`+
 user04052466453699787802                                                                                                        column=family:field0, timestamp=1460701028346, value=')r*>|";21(j92>)['3Nm;%`?&49W3)[c6:4'[w'^35#v&G%'- )!z $ =Ho:Gw;> %Y15E9)*r'O9+0f;W/"/64,v,3$,?<$1z*
 user04052466453699787802                                                                                                        column=family:field1, timestamp=1460701028346, value=9N{:J=9L38!b<J}";d3_y7(03Y#1%t#'b.A%%:l2E}5Zs>1|.5((.02Bc"Ki1T}+V}=76$7v(+ 5 $(Dc?S/0Qc2=z*Oy.H;/O{;
 user04052466453699787802                                                                                                        column=family:field2, timestamp=1460701028346, value="Le:@#;4( ?`4-*$W;?:<4H'7]34G;9Vs#_;8Aw+Q3($h30n.Bm?J+.X%!'~:$h#=*4&65Og+?j+(b!E!8>&%>0(1*9"f&;"#<>!
 user04052466453699787802                                                                                                        column=family:field3, timestamp=1460701028346, value=%Jc7)$#1(%,t5^!58t6A=$Vs!Cc';d//x;Hg.Mq+&j4F\x7F3.2#'(%*$6-l/2*<)"8=r)Co6S/*K#8Wm;Eu7%4:<r20n9(x>;t6Cg7
 user04052466453699787802                                                                                                        column=family:field4, timestamp=1460701028346, value=+0"0U14B\x7F7&t Ty&D+4Cs3Jw2T'=H)">2<Gw9^54Bk!B# /"-*|2#&+Qs2"f49|.6$(\x5C/0%(;+z/Me/$x"1t(I\x7F+E}1!"$6t7M%)
 user04052466453699787802                                                                                                        column=family:field5, timestamp=1460701028346, value=$Lu7,z"Wi#;b32&4\x5Cs"Z#<O+<Ym/0b:Vw"-&;B7)[)/]c74* Zg."v8 r2Zu8&(1Ze-H)-/~'Du+O%34d.6(5Ei.8z$Hc;T9<R?=
 user04052466453699787802                                                                                                        column=family:field6, timestamp=1460701028346, value=)'r7S'(9n&@;'(.87d7W+0O+<%f> n0H50<>*6((52<H\x7F#2v<K?9[w: 04:t<:n+*l5.x6N7:"2)20+Dy+#"-:.%?~.Cs#>j6/24
 user04052466453699787802                                                                                                        column=family:field7, timestamp=1460701028346, value=/@#,<>:!.(V{%%&<Cs-!257d#O'$Ia4@%-\x5Cm2.b.\x5C%61 1'r'!b2>>4"2( <,E51.b3;69E!5/>,0$<N70Qi;'41Ek9).)]=6J/1
 user04052466453699787802                                                                                                        column=family:field8, timestamp=1460701028346, value=#,0<Qu$2<!0."6t.I%$E9-^c!C\x7F+)b$,"=Ci8Qs?6l>Zg:!`4Xo&M;&*n6]#&Q9(0`>$,1I3=*n?52#Ig/?:.5p*".4S718(6V+%
 user04052466453699787802                                                                                                        column=family:field9, timestamp=1460701028346, value=80>$<*&O{.":+Qk?A# 8,53l<M!1,p*Co"R'4Q7)Vy"968.v+&84$l?,*8?d.&29K?=0d3):'Jg+Ee,, <;z,Nw7Oq2Y7)K)2;v+
```

Only odd thing is tabs rather than space so formatting seems off and different formatting from what is in the background code. E.G:

104 -      keychooser = new ScrambledZipfianGenerator(recordcount + expectednewkeys);$
105 +      keychooser=new ScrambledZipfianGenerator(insertstart,insertstart+insertcount+expectednewkeys);$

Notice removal of spaces around '=' and '+'.

Not sure if you lot around here in ycsb are sticklers for this kinda thing or if you'd be up for fixing on commit (remove the tabs).

danielpoltx commented:
Seems like I can't reply to comments directly anymore. This is related to @saintstack patch testing
> I did this and stuff it hung:
It's because your insertstart is bigger than the recordcount. So you're trying to start outside of your record range. I can put a check and exit with an error in that case. Should I ?
> though I'm not sure the zeropadding is working?
you can check zeropadding is working using -p debug=true. In order for it to be 'working' zeropadding number has to be bigger than the number of digits in your key. Try for example something with recordcount=10 and zeropadding>1
> Only odd thing is tabs rather than space so formatting
Well I don't usually code and this was written in Notepad++. It looked fine on my end :)
I'll fix that if I have to commit new changes (convert all tabs to spaces) and add spaces around +, =

saintstack commented:
> Should I ?

Yes. Throw an exception. The hang will confuse folks.

> Well  I don't usually code and this was written in Notepad++

You are doing a good job for someone who doesn't code. I can help w/ the formatting if that helps. Just follow the convention of the background code which seems to put spaces around equals and plus signs.

> Try for example something with recordcount=10 and zeropadding>1

I did but I see no effect. I must be doing something wrong:
```

 $ ./bin/ycsb load hbase10  -P ./workloads/workloadc -p table=ycsb2 -cp ~/conf_hbase -p columnfamily=family -threads 1 -p recordcount=10  -p zeropadding=2


hbase(main):001:0> scan 'ycsb2'
ROW                                                                                                                              COLUMN+CELL
 user1000385178204227360                                                                                                         column=family:field0, timestamp=1460735315380, value=50d7)6?(,1#x"M58=~9@q+"~"#<-Ww6_)5%6?_i3Q7)?*4Qm#'&51f=Zc'6>6)"8I)6?f+=d;+,8\x5C16T+),|+,8!Vo2F54V;:Ek7
 user1000385178204227360                                                                                                         column=family:field1, timestamp=1460735315380, value=56:(^/9F7)Y;'[?1=:,Ec*=*6_5-*>+N9) `.V7,'$ 2|&N?):8= b04l.+:<^c;Eu)(>0R%=,h2G)!T#<'v;[m>3n<H#0Fm)E{;
 user1000385178204227360                                                                                                         column=family:field2, timestamp=1460735315380, value=!X=)30;\x5C}#>~1/>58z>6&$Ka?0z3U= -d>(2;Mm/Z'!+8%[c.L9(By&)b/Fk5,6()"*_a6Mc?;p:Og7.8$^c4Cs;W=<,j.)v,N5,
....
```

danielpoltx commented:
> I did but I see no effect. I must be doing something wrong:

Use insertorder=ordered  otherwise you get hashes which are very long numbers. I suggest you put an operationcount=10 or something like that and debug=true. That way its easier to see what's happening

saintstack commented:
That worked w/ it set to 20 (tried '2' too)... nice.

> user00000000000000000000                                                                        column=family:field0, timestamp=1460751529450, value=*;*,H#1Ne&,j.I+0Aa=\x5Ck'T-"_y;B301x'76*Kw&%<6(f#V-)3,%I{'Yc/(:&'n6*x<9"4"4&'t60v93n6 v86|>V5*0t.!t-Pw#
>  user00000000000000000000                                                                        column=family:field1, timestamp=1460751529450, value=9'b&5>;*x:&j4^#,$&:7|>".79b3U+17r3Y1-( 1,t8U94 |9]k+ 26'|& p0!4-%  Z7==t$?`'8"#\x5Ce?Y3%X#8Rm1X)8842W!=
>  user00000000000000000000                                                                        column=family:field2, timestamp=1460751529450, value=&72-X-'+*->b>N{6=0/V%4O!'/4,/">Gc*O/.Z71=>"8`'Be$R51X\x7F)) !Fc=)8(#$12j*K+?Y{"Mm3+.9M#:U)9Ww!7.51$:Ne'

busbey commented on core/src/main/java/com/yahoo/ycsb/generator/SequentialGenerator.java:
```diff
@@ -0,0 +1,67 @@
+/**                                                                                                                                                                                
+ * Copyright (c) 2010 Yahoo! Inc. All rights reserved.                                                                                                                             
```
this line needs updating. If this is an entirely new set of code, then it should just be 2016 YCSB Contributors. If it's a refactoring, then it should be 2010 Yahoo! Inc., 2016 YCSB Contributors.

busbey commented on core/src/main/java/com/yahoo/ycsb/generator/SequentialGenerator.java:
```diff
@@ -0,0 +1,67 @@
+ * Generates a sequence of integers 0, 1, ...
+ */
+public class SequentialGenerator extends NumberGenerator
+{
```
nit: I think our current style rules (which aren't enforced in core) put opening braces on the end of the previous line.

busbey commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -32,6 +32,7 @@
 import com.yahoo.ycsb.generator.SkewedLatestGenerator;
 import com.yahoo.ycsb.generator.UniformIntegerGenerator;
 import com.yahoo.ycsb.generator.ZipfianGenerator;
+import com.yahoo.ycsb.generator.SequentialGenerator;
```
nit: imports should be lex sorted

busbey commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -61,10 +62,18 @@
+ * YCSB instance (default: 0)
+ * <LI><b>insertcount</b>: for parallel loads and runs, defines the number of records for this
+ * YCSB instance (default: recordcount)
+ * <LI><b>zeropadding</b>: for generating a record sequence compatible with string sort order by
```
should include a warning about insertorder param (maybe put it after that parameter is introduced as well)

danielpoltx replied:
What kind of warning  ?

busbey replied:
I noticed when Stack was testing, the default insert order hashing everything resulted in effectively no zero padding. Since this was surprising to Stack, I'd expect it to be surprising to others as well. A note that ordered inserts are needed for zero padding (similar to what you mentioned to Stack) would preempt the confusion.

danielpoltx replied:
ordered inserts is not needed for zeropadding. Stack just forgot that he was using hashed keys which were longer than the zeropadding value we has using. As an example, his hashed keys are 19 digits, so he needed to use a zeropadding >19 in order to see any change.

busbey replied:
oh, my bad. thanks for the clarification!

saintstack replied:
I was baffled by zeropadding not having an effect even when I set it to 20. I think a sentence that to see an effect, ensure the zeropadding is > keysize would save others my confusion.... just a sentence.

Let me take a look at the rest of the patch....

busbey commented:
great job on getting this contribution together! I added some comments about minor issues. I'll defer to @saintstack on the review of style / functionality.

danielpoltx commented:
I fixed all the comments by @busbey and added the check for valid insertstart + insertcount combination in relation to recordcount.

saintstack commented:
Patch looks good to me otherwise but suggest that if you are going to add comment on zeropadding, that you give the patch another once over; there are irregularities in the formatting but not enough to block the patch... Clean up any you see on the next go round Daniel. Thanks.

danielpoltx commented:
Where can I find the current formatting rules ? I'm willing to make all the appropriate format changes, I just need to know the rules.

saintstack commented on core/src/main/java/com/yahoo/ycsb/Workload.java:
```diff
@@ -64,7 +65,7 @@ public void init(Properties p) throws WorkloadException
       public Object initThread(Properties p, int mythreadid, int threadcount) throws WorkloadException
       {
-	 return null;
+   return null;
```
Did you add a tab here? Or did you remove a space

saintstack commented on core/src/main/java/com/yahoo/ycsb/generator/SequentialGenerator.java:
```diff
@@ -0,0 +1,62 @@
+   * is appropriate for generators that do not return numeric values.
+   */
+  public int nextInt() {
+    int ret = _countstart + counter.getAndIncrement()%_interval;
```
For example, you have white space on either side of the '=' here but not on the either side of the '%' operator. This is a 'nit'. Not important. But you asked (smile)

saintstack commented on core/src/main/java/com/yahoo/ycsb/generator/SequentialGenerator.java:
```diff
@@ -0,0 +1,62 @@
+  }
+  @Override
+  public Number nextValue() {
+    int ret = _countstart + counter.getAndIncrement()%_interval;
```
ditto

saintstack commented on core/src/main/java/com/yahoo/ycsb/generator/SequentialGenerator.java:
```diff
@@ -0,0 +1,62 @@
+  }
+  @Override
+  public Number lastValue() {
+                  return counter.get() + 1;
```
The spacing looks wrong here. Tabs?

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -393,8 +416,18 @@ public void init(Properties p) throws WorkloadException {
 
-    int insertstart =
-        Integer.parseInt(p.getProperty(INSERT_START_PROPERTY, INSERT_START_PROPERTY_DEFAULT));
+  int insertstart =
```
Is the spacing correct here? Do these statements align w/ those above (are there tabs in here?)

risdenk commented:
@danielpoltx the rules are enforced by checkstyle. There isn't a complete listing of all the rules that I'm aware of. Currently there is a checkstyle.xml file in the root directory that configures the rules. You can check the output of checkstyle with something like the following:

```
mvn clean checkstyle:check -pl core -am
```

There are a lot of checkstyle errors currently in core since it is not enforcing yet.

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -393,8 +416,18 @@ public void init(Properties p) throws WorkloadException {
+  int insertstart =
+    Integer.parseInt(p.getProperty(INSERT_START_PROPERTY,INSERT_START_PROPERTY_DEFAULT));
+  int insertcount =
+    Integer.parseInt(p.getProperty(INSERT_COUNT_PROPERTY,String.valueOf(recordcount-insertstart)));
```
You have no space after the ',' on this line.... whereas in rest of code base, there is always a space after a ','

See also the '-' sign... how there are not spaces around it but there are when you look at operators in the background code

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -452,8 +485,10 @@ public void init(Properties p) throws WorkloadException {
     if (requestdistrib.compareTo("uniform") == 0) {
-      keychooser = new UniformIntegerGenerator(0, recordcount - 1);
-    } else if (requestdistrib.compareTo("zipfian") == 0) {
+      keychooser=new UniformIntegerGenerator(insertstart,insertstart+insertcount-1);
```
No spaces around the '=' here or after the ',' or around the '+'

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -452,8 +485,10 @@ public void init(Properties p) throws WorkloadException {
+      keychooser=new UniformIntegerGenerator(insertstart,insertstart+insertcount-1);
+    } else if (requestdistrib.compareTo("sequential")==0) {
+      keychooser=new SequentialGenerator(insertstart,insertstart+insertcount-1);
+  }else if (requestdistrib.compareTo("zipfian") == 0) {
```
No space here and does this align with statements above (tabs?)

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -467,15 +502,15 @@ public void init(Properties p) throws WorkloadException {
       int expectednewkeys = (int) ((opcount) * insertproportion * 2.0); // 2 is fudge factor
 
-      keychooser = new ScrambledZipfianGenerator(recordcount + expectednewkeys);
+      keychooser=new ScrambledZipfianGenerator(insertstart,insertstart+insertcount+expectednewkeys);
```
spaces

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -467,15 +502,15 @@ public void init(Properties p) throws WorkloadException {
       double hotopnfraction =
           Double.parseDouble(p.getProperty(HOTSPOT_OPN_FRACTION, HOTSPOT_OPN_FRACTION_DEFAULT));
-      keychooser = new HotspotIntegerGenerator(0, recordcount - 1, hotsetfraction, hotopnfraction);
+      keychooser = new HotspotIntegerGenerator(insertstart, insertstart+insertcount-1,  hotsetfraction, hotopnfraction);
```
spaces

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -502,7 +537,11 @@ public String buildKeyName(long keynum) {
+  String value = Long.toString(keynum);
+  int fill = zeropadding - value.length();
+  String prekey = "user";
+  for(int i=0; i<fill;i++) {prekey += '0';}
```
spaces

saintstack commented on core/src/main/java/com/yahoo/ycsb/workloads/CoreWorkload.java:
```diff
@@ -502,7 +537,11 @@ public String buildKeyName(long keynum) {
+  int fill = zeropadding - value.length();
+  String prekey = "user";
+  for(int i=0; i<fill;i++) {prekey += '0';}
+  return prekey + value;
```
Is this properly aligned?

danielpoltx commented:
Me and my big mouth. I had to ask :) I'll get on it and fix all the formatting issue with these 3 files, hopefully checkstyle will pass when I'm done

danielpoltx commented:
I  looked at these 3 files and fixed all checkstyle formatting errors. I added additional info on zeropadding parameter. I fixed all formatting suggestions described here.

saintstack commented:
Looks like conflict merging. Did you update before you submitted your patch @danielpoltx ?

Otherwise +1 from me.

Looks like you did fixup of existing tabs and wrapping long lines in Workload.java too. That helps. Usually though, for the future, it is best to keep your changes focused to your affairs so the patch fits the issue. There are few bits of extra spaces on lone lines in a few places but perhaps these can be squashed on commit?

Nits: If you look at the rest of the code, usually there is aline between methods (See SequentialGenerator.java)

Nits: You add space on end of lines in places like here: "...then, if the $" and here "...insertstart + insertcount - 1,  $"

But +1 from me. Try and fix the conflict.

danielpoltx commented:
Conflict fixed and latest formatting comments fixed (hopefully)

busbey commented:
Thanks @danielpoltx and @saintstack!

Future release manager: this change needs to be calledout in release notes.
